fpdf/fpdf.py

```diff
@@ -2528,14 +2509,27 @@ def multi_cell(
                 "Parameter 'w' and 'h' must be numbers, not strings."
                 " You can omit them by passing string content with txt="
             )
-        new_x = XPos.RIGHT
-        new_y = YPos.NEXT
-        if ln == 1:
-            new_x = XPos.LMARGIN
-        elif ln == 2:
-            new_x = XPos.LEFT
-        elif ln == 3:
-            new_y = YPos.TOP
+        if ln is not None:
+            warnings.warn(
+                (
+                    'The parameter "ln" is deprecated.'
```

Comment:
Given how `ln=X` is a lot more shorter than `new_x=XPos.YYY, new_y=YPos.ZZZ`, I suspect that some `fpdf2` users may find the parameters less handy than the original one.
Hence I wonder if we shouldn't allow users to "ignore" this deprecation / replacement.
I'm thinking about 2 strategies:
* ensure this warning is only raised **once**, so that users can safely ignore it without having their terminal / logs being polluted by tens of warning lines
* subtype `DeprecationWarning` in this case, and suggest to users who would like to continue using `ln` to call ` warnings.simplefilter('ignore', LnArgDeprecationWarning)`.

What do you think about those options @gmischler ? Should we adopt both?

Reply:
According to the documentation, the default warnings filter only prints each message once per location it was raised from, so the first point is already a given.

Seeing that the new paramaters are much more intuitive, memorable, and precise, I somehow doubt that people will refuse to use them just to save a few keystrokes once. If you want users to be able to treat different warnings differently, then it might make more sense to establish a more systematic hierarchy, than to single out this specific case.

But generally speaking, don't we want them to actually abandon the deprecated features as soon as possible? Unfortunately, that requires to get on their nerves at least a little bit. If you make it too easy to just ignore specific warnings, then they'll get bitten by them much harder once you actually remove the cruft. Or how long do you plan to keep obsolete APIs around?

Reply:
> According to the documentation, the default warnings filter only prints each message once per location it was raised from, so the first point is already a given.

Nice! I didn't realize that.

> But generally speaking, don't we want them to actually abandon the deprecated features as soon as possible? Unfortunately, that requires to annoy them at least a little bit.

There is a good read about deprecation policies: https://steve-yegge.medium.com/dear-google-cloud-your-deprecation-policy-is-killing-you-ee7525dc05dc

> (quote) Some folks believe that [library] developers should shoulder the costs of compatibility, and others believe that platf[library] users should bear the costs.

What I'd like for `fpdf2` is to provide a clear API for newcomers, but at the same time avoid creating trouble for existing users. I really think we should find a way to introduce clean, clear parameters while limiting any "annoyance" as much as possible 😊

> If you make it too easy to just ignore specific warnings, then they'll get bitten by them much harder once you actually remove the cruft. Or how long do you plan to keep obsolete APIs around?

For the reasons explained in that article, probably for a long time!

But I agree that we don't want to "fall between two stools" here. I think that if I'm reluctant to force users to switch to your new `new_x=XPos.YYY, new_y=YPos.ZZZ` syntax that's simply because I am not fully convinced by it yet...
It is quite verbose to write & read compared to the original `ln=X` system, even if it was very obscure.
But I also see its benefits.

I'm going to leave ourselves some time to think before merging this.

And also, we will need to add some clear documentation about this anyway.

Reply:
That article talks about "platforms" (cloud services, Emacs, Android, Java, web browsers). And in that realm, I mostly agree with the author: Making stuff disappear can be a problem there, primarily because the user has little control over which version they can use and what third-party components it needs to work with.

Fpdf is almost the exact opposite of such a platform, so the same conclusions don't necessarily apply here.
It is a "minimalistic" library, that is usually combined as a small part within a larger software system.
Developers are free to decide if and when they are going to upgrade. And by the very nature of the upgrade concept, they will expect changes when they do that. Anyone who doesn't like a new version for whatever reason, will be able to keep running the old one pretty much forever.

If a developer publishes their stuff eg. via pip, they can (and should!) declare the **tested to be compatible** library versions in the dependency manifest. If a library feature disappears in a future version, the developers who followed that rule won't even notice until they voluntarily upgrade, and their end users will never notice at all.

Due to its history, fpdf has accumulated a lot of technical debt, as a concequence of some suboptimal decisions that were made in the distant past. Parts of its API have been cobbled together rather haphazardly and will eventually need to be fixed.

This "ln" parameter is a prime example for that. It's quite obvious that at first it was simply a 1|0 switch: "should cell() jump to the next line or not." Then somone decided to overload it with additional semantics and also use it to control the horizontal offset. And on top of that, when the concept was transferred to multi_cell(), the meaning of the value "0" was changed to mean "next line" instead of "same line". 

The result makes no sense at all. Even after working on it for a while, trying to replace it as seamlessly as possible, I still can't memorize which value means what. Any sane user should be happy to see that nonsense disappear.

If you want fpdf to remain "minimalistic", then you must get rid of obsolete cruft as soon as humanly possible. This means you need a deprecation strategy **with a timeline**. At the moment, "deprecated" just means "may go away eventually". This seems way too vague, and there's no good reason to be vague here.

"Explicit is better than implicit."
My reccommendation would be, whenever you declare something as deprecated, you also immediately decide for how long it will be staying around, and include that information in the warning message:
```
DeprecationWarning: Use "something" instead. Will be removed in version X.Y.
```
And the closer you get to X.Y, the more annoying you can make that warning. 
Declaring a timeline explicitly like that best satisfies the principle of "least surprise".

I would set the target version to something that you expect to be between 6 and 12 months away (based on the recent update cadence). But you can vary that depending on the severity of the change or your personal taste.